indexing-service/src/main/java/io/druid/indexing/overlord/setup/EqualDistributionWorkerSelectStrategy.java

```diff
@@ -19,41 +19,89 @@
 
 package io.druid.indexing.overlord.setup;
 
-import com.google.common.base.Optional;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
 import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
 import com.google.common.collect.Sets;
 import io.druid.indexing.common.task.Task;
 import io.druid.indexing.overlord.ImmutableWorkerInfo;
 import io.druid.indexing.overlord.config.WorkerTaskRunnerConfig;
 
 import java.util.Comparator;
+import java.util.Map;
+import java.util.Objects;
 import java.util.TreeSet;
 
 /**
  */
 public class EqualDistributionWorkerSelectStrategy implements WorkerSelectStrategy
 {
+  private final AffinityConfig affinityConfig;
+
+  @JsonCreator
+  public EqualDistributionWorkerSelectStrategy(
+      @JsonProperty("affinityConfig") AffinityConfig affinityConfig
+  )
+  {
+    this.affinityConfig = affinityConfig;
+  }
+
+  @JsonProperty
+  public AffinityConfig getAffinityConfig()
+  {
+    return affinityConfig;
+  }
+
   @Override
-  public Optional<ImmutableWorkerInfo> findWorkerForTask(
-      WorkerTaskRunnerConfig config, ImmutableMap<String, ImmutableWorkerInfo> zkWorkers, Task task
+  public ImmutableWorkerInfo findWorkerForTask(
+      final WorkerTaskRunnerConfig config,
+      final ImmutableMap<String, ImmutableWorkerInfo> zkWorkers,
+      final Task task
   )
   {
-    // the version sorting is needed because if the workers have the same available capacity only one of them is
-    // returned. Exists the possibility that this worker is disabled and doesn't have valid version so can't
-    // run new tasks, so in this case the workers are sorted using version to ensure that if exists enable
-    // workers the comparator return one of them.
+    return WorkerSelectUtils.selectWorker(
+        task,
+        zkWorkers,
+        config,
+        affinityConfig,
+        EqualDistributionWorkerSelectStrategy::selectFromEligibleWorkers
+    );
+  }
+
+  private static ImmutableWorkerInfo selectFromEligibleWorkers(final Map<String, ImmutableWorkerInfo> eligibleWorkers)
+  {
     final TreeSet<ImmutableWorkerInfo> sortedWorkers = Sets.newTreeSet(
         Comparator.comparing(ImmutableWorkerInfo::getAvailableCapacity).reversed()
-                  .thenComparing(zkWorker -> zkWorker.getWorker().getVersion()));
-    sortedWorkers.addAll(zkWorkers.values());
-    final String minWorkerVer = config.getMinWorkerVersion();
-
-    for (ImmutableWorkerInfo zkWorker : sortedWorkers) {
-      if (zkWorker.canRunTask(task) && zkWorker.isValidVersion(minWorkerVer)) {
-        return Optional.of(zkWorker);
-      }
+    );
```

Comment:
Same here. Does this TreeSet is needed?

Reply:
Will also change this to a simpler `min` function.